posthog/models/action/action.py

```diff
@@ -42,17 +61,45 @@ def get_analytics_metadata(self):
             "name_length": len(self.name),
             "custom_slack_message_format": self.slack_message_format != "",
             "event_count_precalc": self.events.count(),  # `precalc` because events are computed async
-            "step_count": self.steps.count(),
-            "match_text_count": self.steps.exclude(Q(text="") | Q(text__isnull=True)).count(),
-            "match_href_count": self.steps.exclude(Q(href="") | Q(href__isnull=True)).count(),
-            "match_selector_count": self.steps.exclude(Q(selector="") | Q(selector__isnull=True)).count(),
-            "match_url_count": self.steps.exclude(Q(url="") | Q(url__isnull=True)).count(),
-            "has_properties": self.steps.exclude(properties=[]).exists(),
+            "step_count": len(self.steps),
+            "match_text_count": sum(1 if step.text else 0 for step in self.steps),
+            "match_href_count": sum(1 if step.href else 0 for step in self.steps),
+            "match_selector_count": sum(1 if step.selector else 0 for step in self.steps),
+            "match_url_count": sum(1 if step.url else 0 for step in self.steps),
+            "has_properties": any(step.properties for step in self.steps),
             "deleted": self.deleted,
         }
 
+    @property
+    def steps(self) -> list[ActionStepJSON]:
+        if self.steps_json is None:
+            db_steps = self.action_steps.all()
+            return [
+                ActionStepJSON(
+                    tag_name=step.tag_name,
+                    text=step.text,
+                    text_matching=step.text_matching,
+                    href=step.href,
+                    href_matching=step.href_matching,
+                    selector=step.selector,
+                    url=step.url,
+                    url_matching=step.url_matching,
+                    event=step.event,
+                    properties=step.properties,
+                )
+                for step in db_steps
+            ]
+
+        return [ActionStepJSON(**step) for step in self.steps_json]
+
+    @steps.setter
+    def steps(self, value: list[dict]):
+        # TRICKY: This is a little tricky as DRF will deserialize this here as a dict but typing wise we would expect an ActionStepJSON
+        self.steps_json = [asdict(ActionStepJSON(**step)) for step in value]
+
     def get_step_events(self) -> list[str]:
-        return [action_step.event for action_step in self.steps.all()]
+        # NOTE: This used to return empty strings sometimes - was that on purpose...
```

Comment:
I _think_ an empty string means "All events". @mariusandra Is that right?

Reply:
Yeah the comment was because before the typing was loose which meant that it would return an empty string even though the type said it should always be a list of strings. Not sure if that was on purpose and not sure what should be done here now that we know the type

Reply:
So to be clear I could:
- Cast `action_step.event` to str
- Filter it out (current situation)
- Change the return type to `list[Optional[str]]` but then need to change everywhere it is used to make sure it accounts for this

Reply:
Indeed, an empty string should mean "all events".